lightningd/payalgo.c

```diff
@@ -683,6 +687,7 @@ static void json_pay(struct command *cmd,
 	/* Start with no failures */
 	list_head_init(&pay->pay_failures);
 	pay->in_sendpay = false;
+	pay->description = b11->description;
```

Comment:
We should set pay->description to either `desc` or `b11->description`, one of which must not be NULL.  Unfortunately we'll still have NULL descriptions when we load from the DB, but we can wrap that in COMPAT_V06:

```diff
diff --git a/Makefile b/Makefile
index 6c83a780..57a3131c 100644
--- a/Makefile
+++ b/Makefile
@@ -34,8 +34,8 @@ PIE_LDFLAGS=-pie
 endif
 
 ifeq ($(COMPAT),1)
-# We support compatibility with pre-0.6.
-COMPAT_CFLAGS=-DCOMPAT_V052=1
+# We support compatibility with 0.6, and even before.
+COMPAT_CFLAGS=-DCOMPAT_V052=1 -DCOMPAT_V06=1
 endif
 
 PYTEST_OPTS := -v -x
```

Reply:
The null description is always possible because we can also insert a payment through the `sendpay` command, so I'm happy to keep the conditional display in the json result returned by `listpayments`.

Reply:
Actually not so sure we want to treat the description in the bolt11 the same way we treat the description the `description_hash` commits to. For one we seem to be forcing users to provide a matching description to the hash in order to process a payment at all, this might be a bit cumbersome. Furthermore we currently assume that the description is a UTF-8 string, which might not be the case, e.g., if it commits to a PDF invoice. The spec specifies that the description `d` must be a UTF-8 string, so that is safe, however it does not say that the description hash `h` also commits to a string (which is functionality that I'd like to keep btw).

So simply using `desc` is probably not what we want, since it might not actually be a UTF-8 string, would require hex encoding to be usable in JSON, and wouldn't make much sense to a user in `listpayments` anyway.

Reply:
OK, I'm persuaded.  Deeply tempted to drop this functionality altogether, but not 100% sure we won't want to back, so let's keep.